packages/openspp/src/Adaptor.js

```diff
@@ -0,0 +1,705 @@
+      Log.error(err);
+      throw new Error("Can't login to OpenSPP, please check your credentials or network!");
+    }
+  });
+  return state;
+}
+
+/**
+ * Create a brand new program membership for registrant.
+ * @example
+ *  _creatingEnrolledRegistrantToProgram("IND_Q4VGGZPF", "PROG_2023_00000001")
+ * @private
+ * @param {string} registrant_id - registrant_id of group / individual wanted to unenroll 
+ * @param {string} program_id - program_id of program 
+ */
+function _creatingEnrolledRegistrantToProgram(registrant_id, program_id) {
+  if (sppConnector === null) {
+    login(state);
+  }
+  let registrant;
+  let program;
+  sppConnector
+    .search_read(
+      "res.partner",
+      {
+        domain: [["is_registrant", "=", true], ["registrant_id", "=", registrant_id]],
+        limit: 1,
+        fields: ["id"],
+      },
+      (err, res) => {
+        if (err) {
+          Log.error(err);
+        }
+        registrant = res;
+      }
+    )
+    .then(() =>{
+      sppConnector.search_read(
+        "g2p.program",
+        {
+          domain: [["program_id", "=", program_id]],
+          limit: 1,
+          fields: ["id"],
+        },
+        (err, res) => {
+          if (err) {
+            Log.error(err);
+          }
+          program = res;
+        }
+      )
+    })
+    .then(() =>{
+      if (!registrant || !program) {
+        return Log.error("Registrant or Program not exists!");
+      }
+      let registrantId = registrant[0].id;
+      let programId = program[0].id;
+      sppConnector.create(
+        "g2p.program_membership",
+        {
+          program_id: programId,
+          partner_id: registrantId,
+          state: "enrolled",
+        }
+      );
+    });
+};
+
+/**
+ * get group information from OpenSPP
+ * @public
+ * @example
+ * getGroup("GRP_Q4VGGZPF")
+ * @function
+ * @param {string} registrant_id - The registrant_id of the group
+ * @param {function} callback - An optional callback function
+ * @returns {Operation}
+ */
+export function getGroup(registrant_id, callback=false) {
+  if (sppConnector === null) {
+    login(state);
+  }
+  return state => {
+    let defaultDomain = [
+      ["is_registrant", "=", true],
+      ["is_group", "=", true],
+      ["registrant_id", "=", registrant_id]
+    ];
+    let defaultOrder = "id desc";
+    let defaultFields = [
+      "name", "address", "phone", "kind", "registration_date", "registrant_id"
+    ];
+    let options = {
+      domain: defaultDomain,
+      limit: 1,
+      order: defaultOrder,
+      fields: defaultFields
+    };
+    sppConnector.search_read("res.partner", options, (err, group) => {
+      if (err) {
+        return Log.error(err);
+      }
+      if (!group) {
+        return Log.warn(`Group ${registrant_id} not found!`);
+      }
+      Log.info(`Group ${registrant_id} found!`);
+      let nextState = composeNextState(state, group);
+      if (callback) {
+        return callback(nextState);
+      }
+      return nextState;
```

Comment:
Oh this looks tricky -  this is another asynchronous Javascript problem.

`nextState` won't be returned to the next operation here, because it's _inside_ the callback of `search_read`. So `state.data` won't be passed correctly to the next operation.

I don't have time to make a better suggestion right now (my first thought is to return a Promise). Tomorrow I'll take a look and see what other adaptors are doing in this case and make a suggestion for you.

Reply:
@josephjclark hi, I changed the style from callback to async await, can you re-review it again?